### PR TITLE
_unitary_() init for qasm_output.py

### PR DESCRIPTION
**Pull request by Strilanc**
Fixes https://github.com/quantumlib/Cirq/issues/1139

Derived from https://github.com/quantumlib/Cirq/pull/1145 by @vtomole

### CODE REVIEW

**Comment by Strilanc** [comment]
Good bot :)

**Comment by Strilanc** [comment]
Bad bot :(

File: cirq/circuits/qasm_output.py [hunk 1]
```diff
@@ -52,14 +52,20 @@ def _qasm_(self,
...
     def __repr__(self) -> str:
         return 'cirq.QasmUGate({}, {}, {})'.format(self.lmda,
                                                    self.theta,
                                                    self.phi)
 
+    def _unitary_(self) -> np.ndarray:
+        rz_phi_matrix = protocols.unitary(ops.Rz(self.phi * np.pi))
+        ry_theta_matrix = protocols.unitary(ops.Ry(self.theta * np.pi))
+        rz_lambda_matrix = protocols.unitary(ops.Rz(self.lmda * np.pi))
+        return rz_phi_matrix.dot(ry_theta_matrix.dot(rz_lambda_matrix))
```
Comment (viathor) [inline cirq/circuits/qasm_output.py]: Optional: You can also compute the four elements of this 2x2 matrix using formula (2) in [this paper](https://arxiv.org/abs/1707.03429). Something like this:

```python
pre_z_phase = np.exp(1j * np.pi * self.lmda)
post_z_phase = np.exp(1j * np.pi * self.phi)
y_phase = np.exp(1j * np.pi * np.theta)
c, s = y_phase.real, 1j * y_phase.imag
return np.array([[pre_z_phase.conj() * post_z_phase.conj() * c, ...]])
```

This should be a bit faster. The existing formulas would make a good unit test though.

Comment (Strilanc) [reply cirq/circuits/qasm_output.py]: I cleaned up the code a bit more, but didn't go all the way down to writing out the matrix. (This is not a method on a hot path.) I referenced that paper as the source.

**Comment by Strilanc** [comment]
Google booooootttt...

**Pull request merged**